### PR TITLE
MINIFICPP-1296 - All tests should use volatile state storage

### PR DESCRIPTION
**Pull request by adam-markovics**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.h [hunk 1]
```diff
@@ -29,19 +29,21 @@
...
 
 namespace org {
 namespace apache {
 namespace nifi {
 namespace minifi {
 namespace controllers {
 
-class UnorderedMapKeyValueStoreService : virtual public KeyValueStoreService {
+/// Key-value store serice purely in RAM without disk usage
+class UnorderedMapKeyValueStoreService : virtual public PersistableKeyValueStoreService {
```
Comment (szaszm) [inline extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.h:41]: If it's RAM only, then why is it `PersistableKeyValueStoreService`?

File: extensions/standard-processors/controllers/UnorderedMapPersistableKeyValueStoreService.h [hunk 2]
```diff
@@ -59,20 +61,30 @@ class UnorderedMapPersistableKeyValueStoreService : public AbstractAutoPersistin
...
   std::string file_;
 
   bool load();
 
-  std::string escape(const std::string& str);
   bool parseLine(const std::string& line, std::string& key, std::string& value);
 
  private:
+  void initializeNonVirtual();
+  bool persistNonVirtual();
```
Comment (szaszm) [inline extensions/standard-processors/controllers/UnorderedMapPersistableKeyValueStoreService.h:86]: Why are these necessary? If subclasses need to call the base class implementation, they can always qualify the call with the base class name to avoid dynamic dispatching,

**Review by szaszm — COMMENTED** [review]
just a cursory first pass

File: libminifi/include/utils/ThreadPool.h [hunk 3]
```diff
@@ -227,11 +227,10 @@ class ThreadPool {
    * Returns true if a task is running.
    */
   bool isTaskRunning(const std::string &identifier) const {
-    try {
-      return task_status_.at(identifier) == true;
-    } catch (const std::out_of_range &) {
+    const auto iter = task_status_.find(identifier);
```
Comment (adamdebreceni) [inline libminifi/include/utils/ThreadPool.h:230]: I know it's old code but I wonder why aren't we locking the `worker_queue_mutex_` or the `manager_mutex_` mutex here, even though we lock at least one of them every other place we access the `task_status_` map

File: extensions/standard-processors/processors/TailFile.cpp [hunk 4]
```diff
@@ -497,11 +497,8 @@ bool TailFile::recoverState(const std::shared_ptr<core::ProcessContext>& context
...
   if (state_manager_->get(state_map)) {
-    for (size_t i = 0U;; i++) {
-      std::string name;
-      try {
-        name = state_map.at("file." + std::to_string(i) + ".name");
-      } catch (...) {
+    for (size_t i = 0U;; ++i) {
+      if (state_map.find("file." + std::to_string(i) + ".name") == state_map.end()) {
         break;
       }
```
Comment (adamdebreceni) [inline extensions/standard-processors/processors/TailFile.cpp:503]: I am genuinely curious what was the reasoning behind this code, could we be missing something by rewriting it to `std::map<...>::find`?

File: extensions/standard-processors/tests/unit/TailFileTests.cpp [hunk 5]
```diff
@@ -1129,8 +1129,11 @@ TEST_CASE("TailFile finds and finishes the renamed file and continues with the n
   utils::Identifier tail_file_uuid = utils::IdGenerator::getIdGenerator()->generate();
   const core::Relationship success_relationship{"success", "everything is fine"};
 
+  // use persistent state storage that defaults to rocksDB, not volatile
+  const auto configuration = std::make_shared<minifi::Configure>();
```
Comment (adamdebreceni) [inline extensions/standard-processors/tests/unit/TailFileTests.cpp:1133]: is this necessary? `TestController::createPlan` default constructs a `Configure` object if `nullptr` is provided, and by default a `RocksDbPersistableKeyValueStoreService` is constructed

File: extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.h (hunk 1 above)
Comment (adam-markovics) [reply extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.h:41]: Because AbstractCoreComponentStateManagerProvider (grandparent of this class) declares a pure virtual persistImpl() method that is implemented in base class PersistableKeyValueStoreService by calling persist(). Basically, persist() is a hook, that can be called with any implementation, doing a null operation when it is volatile.

File: extensions/standard-processors/controllers/UnorderedMapPersistableKeyValueStoreService.h (hunk 2 above)
Comment (adam-markovics) [reply extensions/standard-processors/controllers/UnorderedMapPersistableKeyValueStoreService.h:86]: Because initialize() and persist() were called in constructor/destructor and they are virtual methods.

File: libminifi/include/utils/ThreadPool.h (hunk 3 above)
Comment (adam-markovics) [reply libminifi/include/utils/ThreadPool.h:230]: Good question, I added locking `worker_queue_mutex_` as it is locked by most other locations of accessing `task_status_`, and its documentation comment says `mutex to  protect task status and delayed queue`.

File: extensions/standard-processors/processors/TailFile.cpp (hunk 4 above)
Comment (adam-markovics) [reply extensions/standard-processors/processors/TailFile.cpp:503]: I don't think we would miss anything. `std::unordered_map<...>::at()` only throws if the requested element is not found, the new code is also checking that. My goal was to eliminate exceptions as they are confusing during debugging and should not be used for control flow. And especially eliminating `catch (...)` from wherever possible.

File: extensions/standard-processors/tests/unit/TailFileTests.cpp (hunk 5 above)
Comment (adam-markovics) [reply extensions/standard-processors/tests/unit/TailFileTests.cpp:1133]: No, UnorderedMapKeyValueStoreService will be the default from now on for nullptr:
https://github.com/apache/nifi-minifi-cpp/pull/920/files/74dc246506795eb3832547da796c227d03ecb321#diff-530e3557b16fbcc2c012a6ac10efccf62258f51004a7c61d3191334cbd3e3461
In this test case, persistency is more of a requirement, so volatile is not sufficient.

File: libminifi/test/TestBase.h [hunk 6]
```diff
@@ -282,18 +274,10 @@ class TestPlan {
...
-  }
-
   std::shared_ptr<logging::Logger> getLogger() const {
     return logger_;
   }
 
-  std::string getStateDir() {
-    return state_dir_;
-  }
-
```
Comment (szaszm) [inline libminifi/test/TestBase.h:296]:
```
/home/szaszm/nifi-minifi-cpp-2/extensions/sftp/tests/ListSFTPTests.cpp: In member function ‘void ListSFTPTestsFixture::createPlan(org::apache::nifi::minifi::utils::Identifier*)’:
/home/szaszm/nifi-minifi-cpp-2/extensions/sftp/tests/ListSFTPTests.cpp:92:64: error: ‘using element_type = class TestPlan’ {aka ‘class TestPlan’} has no member named ‘getStateDir’
92 |     const std::string state_dir = plan == nullptr ? "" : plan->getStateDir();
|                                                                ^~~~~~~~~~~
make[2]: *** [extensions/sftp/tests/CMakeFiles/ListSFTPTests.dir/build.make:82: extensions/sftp/tests/CMakeFiles/ListSFTPTests.dir/ListSFTPTests.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:8318: extensions/sftp/tests/CMakeFiles/ListSFTPTests.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Review by szaszm — CHANGES_REQUESTED** [review]

**Review by szaszm — APPROVED** [review]

**Review by adamdebreceni — APPROVED** [review]

**Comment by adamdebreceni** [comment]
the `ListSFTPTests` failed on `ubuntu-16.04-all` on all runs I saw, (here in the PR, and on your own fork run), could we re-run it again (preferably on both places) to verify if the failures are only transient?

**Comment by szaszm** [comment]
ListSFTPTests is still failing after another rerun, so this needs some attention.

**Comment by szaszm** [comment]
35999e0 (on main) broke the build
